nextcord/cooldowns/buckets.py

```diff
@@ -0,0 +1,59 @@
+from enum import Enum
+
+
+class _HashableArguments:
+    # An internal class defined such that we can
+    # use *args and **kwargs as keys. We need to
+    # do this as mutable items are not hashable,
+    # therefore are not suitable for usage as
+    # dictionary keys. Thus this wraps those
+    # arguments with repr hashes used
+    def __init__(self, *args, **kwargs):
+        self.args: tuple = args
+        self.kwargs: dict = kwargs
+
+    def __repr__(self):
+        return f"_HashableArguments(args={self.args}, kwargs={self.kwargs})"
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, type(self)):
+            return False
+
+        return self.args == other.args and self.kwargs == other.kwargs
+
+    def __hash__(self) -> int:
+        # TODO This needs to be better suited to uniqueness and nesting
+        return hash(repr(self.args) + repr(self.kwargs))
+
+
+class CooldownBucket(Enum):
```

Comment:
This should probably be more descriptivly named like FunctionCooldownBucket or similar?

Reply:
N/A as it will cover everything